cmd/agent/app/reporter/connect_metrics.go

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) 2020 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package reporter
+
+import (
+	"time"
+
+	"github.com/uber/jaeger-lib/metrics"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+)
+
+// Structure built for future code expansion or add new metrics.
+type connectMetrics struct {
+}
+
+// ConnectMetricsReporterParams is used as input to WrapWithConnectMetrics.
+type ConnectMetricsReporterParams struct {
+	Logger          *zap.Logger     // required
+	MetricsFactory  metrics.Factory // required
+	ExpireFrequency time.Duration
+	ExpireTTL       time.Duration
+}
+
+// ConnectMetricsReporter is a decorator also it not actual use currently.
+// Structure built for future code expansion
+type ConnectMetricsReporter struct {
+	params         ConnectMetricsReporterParams
+	connectMetrics *connectMetrics
+	shutdown       chan struct{}
+	closed         *atomic.Bool
+}
+
+// WrapWithConnectMetrics creates ConnectMetricsReporter.
+func WrapWithConnectMetrics(params ConnectMetricsReporterParams) *ConnectMetricsReporter {
+	if params.ExpireFrequency == 0 {
+		params.ExpireFrequency = defaultExpireFrequency
+	}
+	if params.ExpireTTL == 0 {
+		params.ExpireTTL = defaultExpireTTL
+	}
+	cm := new(connectMetrics)
+	params.MetricsFactory = params.MetricsFactory.Namespace(metrics.NSOptions{Name: "connection_status"})
+	metrics.MustInit(cm, params.MetricsFactory, nil)
+	r := &ConnectMetricsReporter{
+		params:         params,
+		connectMetrics: cm,
+		shutdown:       make(chan struct{}),
+		closed:         atomic.NewBool(false),
+	}
+	return r
+}
+
+// CollectorConnected used for change metric as agent connected.
+func (r *ConnectMetricsReporter) CollectorConnected(target string) {
+	metric := r.params.MetricsFactory.Gauge(metrics.Options{
+		Name: "connected_collector_status",
+		Help: "Connection status that jaeger-agent to jaeger-collector, 1 is connected, 0 is disconnected",
+		Tags: map[string]string{"target": target},
```

Comment:
The `target` here is the address of the collector, correct? Is it a good idea to use that as label, given that it could contain an ephemeral port number and prone to high cardinality? I'd say the target could be reported as an expvar variable, but the metric can be just the status.

Reply:
> ephemeral port number

Care to expand on that? I don't remember seeing servers with ephemeral port numbers.

Reply:
Why not? For example, at Uber most services run on ephemeral port numbers, to make it easy to stack and relocate them across available compute capacity. The discovery service reports host:port pairs for all instances of a given service. The custom build of the agent that Uber runs would use that discovery mechanism to locate Jaeger collectors. So the `target` string could be quite dynamic.

Reply:
Actually, for this early version, I use such as `oldTarget` variable, the function incoming variable like `newTarget`, function will change `oldTarget` metric to 0, change `newTarget` metric to 1. In this case, we can watch which one agent connects to which collector. But eventually, I thought these are too complicated. We only need to observe the connection status is ok, after all. Which one agent connects to which collector should be added in collector metrics.

Reply:
For `connected is 0 disconnected is 1` , because when have many connection metrics, we can filter `sum(jaeger_agent_connection_status_connected_collector_status{})  by (instance) > bool  0`  for quick view status.

I fixed the target, no need to pass the target parameter, in the next commit. target is static when the `connectMetrics{}` initialize

For the current PR, I thought that simplify the design first, and add new features if required.

Reply:
> For example, at Uber most services run on ephemeral port numbers, to make it easy to stack and relocate them across available compute capacity

Didn't know that, but it does make sense for high density VMs/bare metal servers.